hnn_core/extracellular.py

```diff
@@ -0,0 +1,512 @@
+                num = np.sqrt(L ** 2 + R2) + L
+                denom = np.sqrt(H ** 2 + R2) + H  # == norm(b) + H
+
+            phi[idx] = np.log(num / denom) / norm_a
+
+    # [dis]: um; [sigma]: S / m
+    # [phi / sigma] = [1/dis] / [sigma] = 1 / [dis] x [sigma]
+    # [dis] x [sigma] = um x (S / m) = 1e-6 S
+    # transmembrane current returned by _ref_i_membrane_ is in [nA]
+    # ==> 1e-9 A x (1 / 1e-6 S) = 1e-3 V = mV
+    # ===> multiply by 1e3 to get uV
+    return 1000.0 * phi / (4.0 * np.pi * sigma)
+
+
+class ExtracellularArray:
+    """Class for recording extracellular potential fields with electrode array
+
+    Note that to add an electrode array to a simulation, you should use the
+    :meth:`hnn_core.Network.add_electrode_array`-method. After simulation,
+    the network will contain a dictionary of `ExtracellularArray`-objects
+    in ``net.rec_array`` (each array must be added with a unique name). An
+    `ExtracellularArray` contains the voltages at each electrode contact,
+    along with the time points at which the voltages were sampled.
+
+    Parameters
+    ----------
+    positions : tuple | list of tuple
+        The (x, y, z) coordinates (in um) of the extracellular electrodes.
+    sigma : float
+        Extracellular conductivity, in S/m, of the assumed infinite,
+        homogeneous volume conductor that the cell and electrode are in.
+    method : str
+        Approximation to use. 'psa' (default) assigns all transmembrane
+        currents to the center point (0.5) (point source approximation).
+        'lsa' treats the section as a line source, and a single multiplier is
+        calculated for each section (line source approximation).
+    times : None | list of float
+        Optionally, provide precomputed voltage sampling times for electrodes
+        at `positions`.
+    voltages : None | list of list of list of float (3D)
+        Optionally, provide precomputed voltages for electrodes at `positions`.
+        Note that the size of `voltages` must be: (n_trials, n_electrodes,
+        n_times), i.e., three-dimensional.
+
+    Attributes
+    ----------
+    times : list of float
+        The time points the extracellular voltages are sampled at (ms)
+    sfreq : float
+        Sampling rate of the extracellular data (Hz).
+
+    Notes
+    -----
+    See Table 5 in http://jn.physiology.org/content/104/6/3388.long for
+    measured values of sigma in rat cortex (note units there are mS/cm)
+    """
+
+    def __init__(self, positions, sigma=0.3, method='psa', times=None,
+                 voltages=None):
+        _validate_type(positions, (tuple, list), 'positions')
+        if isinstance(positions, tuple):
+            positions = [positions]
+        for pos in positions:
+            if len(pos) != 3:
+                raise ValueError('positions should be provided as xyz '
+                                 f'coordinate triplets, got: {positions}')
+        _validate_type(sigma, float, 'sigma')
```

Comment:
A lot of these input checks are redundant with `net.add_electrode_array`. Perhaps they can be moved to a `validate_lfp_parameters(positions, sigma, ...)` function that lives in this file, and is imported in `network.py`.